File: Frontend/Swift/barbicha/Model/Barbershop.swift
//
//  Barbershop.swift
//  barbicha
//
//  Created by Gustavo Tavares on 16/04/2018.
//  Copyright © 2018 Gustavo Tavares. All rights reserved.
//

import Foundation
import PlazazCore
import SipHash

class Barbershop: PlazazOrganization, Codable {

    public private (set) var uuid: String
    public var name: String?
    public var imageUrl: URL?
    public var imageData: Data?
    private var latitude: Double?
    private var longitude: Double?
    public private (set) var serviceTypes: [AppointmentType]
    public private (set) var barbers: [Barber]
    private var appointments: [Appointment]
    
    public var location: (Double, Double)? {
        
        get {return (self.latitude == nil || self.longitude == nil ? nil : (self.latitude!, self.longitude!))}
        
        set {
            
            if newValue == nil {self.latitude = nil; self.longitude = nil
            } else {self.latitude = newValue!.0; self.longitude = newValue!.1}
            
        }
    }
    
    init(name: String, imageUrl: URL?, services: [AppointmentType]) {
        
        self.uuid = PlazazCoreHelpers.newUUID(for: Barbershop.self)
        self.name = name
        self.imageUrl = imageUrl
        self.serviceTypes = services
        self.appointments = []
        self.barbers = []
        
        if imageUrl != nil {
            
            AppDelegate.imageFromUrl(url: imageUrl!) { (data, error) in
                
                guard error == nil else {return}
                self.imageData = data
                self.signalChange(event: .imageUpdated)
                
            }
            
        } else {self.imageData = AppUtilities.shared.fallbackFromCache(name: uuid, extension: defaultImageFormat)}
        
    }

    public func updateFromCloud(dict: Dictionary<String, Any>) -> Void {
        
        let _name = dict["name"] as? String
        let _imageUrl = dict["imageUrl"] as? String
        let _latitude = dict["latitude"] as? Double
        let _longitude = dict["longitude"] as? Double
        
        var changed = false
        if self.name != _name {self.name = _name; changed = true}
        if self.latitude != _latitude {self.latitude = _latitude; changed = true}
        if self.longitude != _longitude {self.longitude = _longitude; changed = true}
        if self.imageUrl?.absoluteString != _imageUrl {
            
            let newUrl:URL? = _imageUrl == nil ? nil : URL(string: _imageUrl!)
            self.imageUrl = newUrl
            
        }
        
        if changed {self.signalChange(event: .shopUpdated)}
        
    }
    
    public func updateBarbersFromCloud(barbers fromWeb: [Barber]) -> Void {

        var changed: Bool = false
        
        let oldUUIDs:Set<String> = Set(self.barbers.compactMap { $0.uuid })
        let newUUIDs:Set<String> = Set(fromWeb.compactMap { $0.uuid })
        
        let toBeRemoved = oldUUIDs.subtracting(newUUIDs)
        if toBeRemoved.count > 0 {changed = true}
        
        let toBeAdded = newUUIDs.subtracting(oldUUIDs)
        if toBeAdded.count > 0 {changed = true}
        
        let toBeUpdated = oldUUIDs.intersection(newUUIDs)
        
        var updatedArray: [Barber] = []
        for id in toBeUpdated {
            
            let old = self.barbers.filter({ id == $0.uuid }).first
            let new = fromWeb.filter({ id == $0.uuid }).first
            
            if old != nil {
                
                if new != nil {
                    if old != new {
                        changed = true
                        old?.updateValues(with: new!)}}
                
                updatedArray.append(old!)
                
            }

        }
        
        let newArray = fromWeb.filter { toBeAdded.contains($0.uuid) }
        
        if changed {
            self.barbers = newArray + updatedArray
            self.signalChange(event: .barberListUpdated)
        }

    }
    
    public func updateServicesFromCloud(services: [AppointmentType]) -> Void {
        
        self.serviceTypes = services
        
    }
    
    public func updateAppointmentsFromCloud(appointments fromWeb: [Appointment]) -> Void {

        var changed: Bool = false
        
        let oldUUIDs:Set<String> = Set(self.appointments.compactMap { $0.uuid })
        let newUUIDs:Set<String> = Set(fromWeb.compactMap { $0.uuid })
        
        let toBeRemoved = oldUUIDs.subtracting(newUUIDs)
        if toBeRemoved.count > 0 {changed = true}
        
        let toBeAdded = newUUIDs.subtracting(oldUUIDs)
        if toBeAdded.count > 0 {changed = true}
        
        let toBeUpdated = oldUUIDs.intersection(newUUIDs)
        
        var updatedArray: [Appointment] = []
        for id in toBeUpdated {
            
            let old = self.appointments.filter({ id == $0.uuid }).first
            let new = fromWeb.filter({ id == $0.uuid }).first
            
            if old != nil {
                
                if new != nil {
                    if old != new {
                        changed = true
                        old?.updateValues(with: new!)}}
                
                updatedArray.append(old!)
                
            }
            
        }
        
        let newArray = fromWeb.filter { toBeAdded.contains($0.uuid) }
        
        if changed {
            self.appointments = newArray + updatedArray
            self.signalChange(event: .appointmentListUpdated)
        }
        
        
        
    }

    public func addBarber(_ barber: Barber) -> Void {
    
        self.barbers.append(barber)
        self.signalChange(event: .barberListUpdated)
        
    }
    
    public func removeBarber(_ barber: Barber) -> Void {
        
        var i: Int = 0
        var toBeRemoved: Int? = nil
        for item in self.barbers {
            if item == barber {toBeRemoved = i}
            i = i + 1
        }
        
        if let found = toBeRemoved {
            self.barbers.remove(at: found)
            self.signalChange(event: .barberListUpdated)
        }

    }
    
    public func appointments(for date: Date, with barberId: String?) -> [Appointment] {

        var appointmentList: [Appointment] = []
        
        let startTime:Date = self.startTime(for: date)
        let endTime:Date = self.endTime(for: date)
        let slot:Int = self.slotTime(for: date)
        
        var dateWalk:Date = startTime
        while dateWalk <  endTime {

            let comparationStart = dateWalk.addingTimeInterval(-10)
            let comparationEnd = dateWalk.addingTimeInterval(10)
            
            var found = self.appointments.filter {
                
                let fromBarber = barberId == nil ? true : barberId! == $0.barberUUID!
                return $0.startDate > comparationStart && $0.startDate < comparationEnd && fromBarber
                
            }
            
            let value = found.count > 0 ? found[0] : Appointment.empty(for: dateWalk, interval: slot)

            appointmentList.append(value)
            dateWalk.addTimeInterval(Double(slot * 60))
            
        }
        
        return appointmentList
        
    }

    public func confirm(for date: Date, type: AppointmentType, customer: PlazazPerson, handler: @escaping (Bool) -> ()) -> Void {}
    
    public func cancel(for date: Date, type: AppointmentType, customer: PlazazPerson, handler: @escaping (Bool) -> ()) -> Void {}
    
    public func move(for date: Date, type: AppointmentType, customer: PlazazPerson, handler: @escaping (Bool) -> ()) -> Void {}
    
}

File: Frontend/Android/app/src/main/java/br/com/plazaz/barbicha/model/Barbershop.kt
package br.com.plazaz.barbicha.model

import android.util.Log
import br.com.plazaz.barbicha.logTag
import br.com.plazaz.barbicha.protocols.*
import java.net.URL
import java.util.*
import kotlin.collections.ArrayList

class Barbershop(uuid: String, name: String, imageURL: URL? = null, latitude: Double? = null, longitude: Double? = null,
                 serviceTypes: ArrayList<AppointmentType> = arrayListOf(), barbers: ArrayList<Barber> = arrayListOf(), appointments: ArrayList<Appointment> = arrayListOf()) : EventEmitter {

    val uuid: String = uuid;
    var name: String = name;
    var latitude: Double? = latitude;
    var longitude: Double? = longitude;
    var imageURL: URL? = imageURL;
    var imageData: ByteArray? = null;

    var serviceTypes: ArrayList<AppointmentType> = serviceTypes
        private set;

    var barbers: ArrayList<Barber> = barbers
        private set;

    var appointments: List<Appointment> = appointments
        private set;

    var observers: ArrayList<EventObserver> = arrayListOf()

    fun location(): Pair<Double?, Double?> = Pair(this.latitude, this.longitude)

    fun updateFromCloud(map: Map<String, Any>) {

        val name = map.get(kName) as String
        val lat = map.get(kLatitude) as Double?
        val lon = map.get(kLongitude) as Double?
        val image = map.get(kImageUrl) as? String
        val imageUrl = if (image != null) URL(image) else null

        var changed: Boolean = false

        if (this.name == name) {this.name = name; changed = true}
        if (this.latitude == lat) {this.latitude = lat; changed = true}
        if (this.longitude == lon) {this.longitude = lon; changed = true}
        if (this.imageURL == imageUrl) {this.imageURL = imageUrl; changed = true}

        if (changed) {this.informEvent(Event(Barbershop.kEventBarbershopUpdated, arrayListOf(this.uuid)))}

    }

    fun updateBarbersFromCloud(fromWeb: Set<Barber>) {

        var changed: Boolean = false

        val oldUUIDs:Set<String> = this.barbers.map { it.uuid }.toSet()
        val newUUIDs:Set<String> = fromWeb.map { it.uuid }.toSet()

        val toBeRemoved = oldUUIDs.minus(newUUIDs)
        if (toBeRemoved.isNotEmpty()) {changed = true}

        val toBeAdded = newUUIDs.minus(oldUUIDs);
        if (toBeAdded.isNotEmpty()) {changed = true}

        val toBeUpdated = oldUUIDs.intersect(newUUIDs);

        var updatedSet: Array<Barber> = emptyArray();
        for (id in toBeUpdated) {

            val old: Barber? = this.barbers.filter({ id == it.uuid }).first()
            val new: Barber? = fromWeb.filter({ id == it.uuid }).first()

            if (old != null) {

                if (new != null) {
                    if (old != new) {
                        changed = true
                        old.updateValues(new)}}

                updatedSet = updatedSet.plus(old);

            }

        }

        var newSet = fromWeb.toSet().filter { toBeAdded.contains(it.uuid) }

        if (changed) {
            this.barbers = ArrayList(newSet.plus(updatedSet));
            this.informEvent(Event(Barbershop.kEventBarberListUpdated, arrayListOf(this.uuid)))
        }

    }

    fun updateServicesFromCloud(services: ArrayList<AppointmentType>) {

        this.serviceTypes = services
        this.informEvent(Event(Barbershop.kEventBarbershopUpdated, arrayListOf(this.uuid)))

    }

    fun updateAppointmentsFromCloud(fromWeb: ArrayList<Appointment>) {

        var changed: Boolean = false

        val oldUUIDs:Set<String> = this.appointments.map { it.uuid }.toSet();
        val newUUIDs:Set<String> = fromWeb.map { it.uuid }.toSet();

//        val toBeRemoved = oldUUIDs.minus(newUUIDs)
//        if (toBeRemoved.isNotEmpty()) {changed = true}

        val toBeAdded = newUUIDs.minus(oldUUIDs)
        if (toBeAdded.isNotEmpty()) {changed = true}

        val toBeUpdated = oldUUIDs.intersect(newUUIDs)

        for (id in toBeUpdated) {

            val old: Appointment? = this.appointments.filter({ id == it.uuid }).first()
            val new: Appointment? = fromWeb.filter({ id == it.uuid }).first()

            if (old != null) {

                if (new != null) {
                    if (old != new) {
                        changed = true
                        old.updateValues(new)}}

            }

        }

        val newList = fromWeb.filter { toBeAdded.contains(it.uuid) }

        if (changed) {
            Log.d(logTag, "Barbershop:updateAppointmentsFromCloud:Old List (${this.appointments.map { it.uuid }})")
            this.appointments = this.appointments.plus(newList)
            Log.d(logTag, "Barbershop:updateAppointmentsFromCloud:New List (${this.appointments.map { it.uuid }})")
            this.informEvent(Event(Barbershop.kEventAppointmentListUpdated, arrayListOf(this.uuid)))

        }

    }

    fun dayStart(date: Date): Date {

        var startTime = Calendar.getInstance(TimeZone.getDefault(), Locale.getDefault())
        startTime.time = date
        startTime.set(Calendar.HOUR_OF_DAY, 9)
        startTime.set(Calendar.MINUTE, 0)
        startTime.set(Calendar.SECOND, 0)
        return startTime.time

    }

    fun dayEnd(date: Date): Date {

        var endTime = Calendar.getInstance(TimeZone.getDefault(), Locale.getDefault())
        endTime.time = date
        endTime.set(Calendar.HOUR_OF_DAY, 21)
        endTime.set(Calendar.MINUTE, 0)
        endTime.set(Calendar.SECOND, 0)
        return endTime.time

    }

    fun startTime(date: Date): Date {

        var startTime = Calendar.getInstance(TimeZone.getDefault(), Locale.getDefault())
        startTime.time = date
        startTime.set(Calendar.HOUR_OF_DAY, 9)
        startTime.set(Calendar.MINUTE, 0)
        startTime.set(Calendar.SECOND, 0)
        return startTime.time

    }

    fun endTime(date: Date): Date {

        var endTime = Calendar.getInstance(TimeZone.getDefault(), Locale.getDefault())
        endTime.time = date
        endTime.set(Calendar.HOUR_OF_DAY, 21)
        endTime.set(Calendar.MINUTE, 0)
        endTime.set(Calendar.SECOND, 0)
        return endTime.time

    }

    fun slotTime(date: Date): Int {
        return 30
    }

    fun getBarberById(barberId: String?): Barber? {

        return this.barbers.filter { it.uuid == barberId }.first()

    }

    fun getAppointmentsFor(date: Date, barberUUID: String?): ArrayList<Appointment> {

        var appointmentList: ArrayList<Appointment> = arrayListOf()

        val startTime:Date = this.startTime(date)
        val endTime:Date = this.endTime(date)
        val slot:Int = this.slotTime(date)

        Log.d(logTag, "Barbershop:getAppointmentsFor:$startTime and $endTime")

        var dateWalkInMillis = startTime
        while (dateWalkInMillis <  endTime) {

            var found = this.appointments.filter<Appointment> { it: Appointment ->

                val fromBarber = if (barberUUID == null) {true} else {barberUUID == it.barberUUID}
                this.compareDatesWithoutMilliseconds(it.startDate, dateWalkInMillis) && fromBarber

            }

            val value:Appointment = if (found.isEmpty()) {Appointment.empty(dateWalkInMillis, slot)} else {found.first()}

            appointmentList.add(value)
            dateWalkInMillis = Date(dateWalkInMillis.time + (1000 * 60 * value.interval).toLong())

        }

        return appointmentList

    }

    override fun registerObserver(observer: EventObserver) {this.observers.add(observer)}

    override fun unregister(observer: EventObserver) {this.observers.remove(observer)}

    override fun informEvent(event: Event) {

        for (observer in this.observers) {observer.receiveEvent(event)}

    }

    private fun compareDatesWithoutMilliseconds(lh: Date, rh: Date): Boolean {return lh.time / 1000 == rh.time / 1000}

    companion object {

        val kUUID: String = "uuid";
        val kName: String = "name"
        val kLatitude: String = "latitude"
        val kLongitude: String = "longitude"
        val kImageUrl: String = "imageUrl"

        val kEventBarbershopUpdated: String = "barbershopUpdated"
        val kEventBarberListUpdated: String = "barberListUpdated"
        val kEventAppointmentUpdated: String = "appointmentUpdated"
        val kEventAppointmentListUpdated: String = "appointmentListUpdated"

        fun fromMap(map: Map<String, Any>): Barbershop {

            val uuid = map.get(kUUID) as String
            val name = map.get(kName) as String
            val lat = map.get(kLatitude) as Double?
            val lon = map.get(kLongitude) as Double?
            val image = map.get(kImageUrl) as? String
            val imageUrl = if (image != null) URL(image) else null

            return Barbershop(uuid, name, imageUrl, lat, lon)
        }

    }

}